docs/design/plan-modification.md

```diff
@@ -0,0 +1,298 @@
+
+// GetSchema returns the schema for this resource.
+func (f fileResourceType) GetSchema(_ context.Context) (schema.Schema, []*tfprotov6.Diagnostic) {
+	return schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"favorite_number": {
+				Type:     types.NumberType,
+				Required: true,
+				RequiresReplaceIf: func(ctx context.Context, old, new attr.Value) bool {
+				  oldVal := old.(types.Number)
+				  newVal := new.(types.Number)
+				  
+				  if !oldVal.Unknown && !oldVal.Null && !newVal.Unknown && !newVal.Null {
+				    if newVal.Value.Cmp(oldVal.Value) > 0 {
+				      return true
+				    }
+				  }
+				  return false
+				},
+			},
+		},
+	}, nil
+}
+```
+
+While this provider implementation of the RequiresReplace function is unfortunately more verbose than its `helper/schema` equivalent, since any value comparison must be preceded by a check for null and unknown values, we can easily imagine a `RequiresReplaceIfSet` helper implemented in the framework, illustrated in the _Composition and other helpers_ section below. The ability to work with null and unknown values is a new feature not available in `helper/schema`.
+
+Since `RequiresReplaceIf` must be implemented on the resource type, not the resource instance, it has no access to the provider's API client. It is intended to be used for the simple old and new value comparisons that make up the majority of observed usages of `CustomizeDiff` in current provider code.
+
+#### Tradeoffs
+
+With easily mocked arguments, `RequiresReplaceIf` functions are unit testable, and, being a field on the `schema.Attribute` struct, easily discoverable. The provider code is reasonably Go-native, as far as the complexities of the various type systems involved allow.
+
+As indicated above, helper functions can be introduced to reduce verbosity.
+
+Compatibility may be an issue if a common use case emerges for RequiresReplace conditions a little more complex than the signature of `RequiresReplaceIfFunc` allows - for example, requiring reading config values. This could be anticipated by having the signature instead be:
+```go
+type RequiresReplaceIfFunc (context.Context, ModifyPlanRequest) bool
+```
+By doing this, however, we would lose the benefits of the `old, new` parameters in reducing verbosity - the provider code would have to repeat the attribute path in order to retrieve the values from `req.Plan` and `req.State`.
+
+### `attr.TypeWithRequiresReplace`
+
+A `ModifyPlan()` or `RequiresReplaceIf()` function could be added to an extension of the `attr.Type` interface:
+
+```go
+type TypeWithRequiresReplace interface {
+  Type
+  
+  RequiresReplaceIf(context.Context, old, new Value) bool
+```
+
+This would allow bundling reusable RequiresReplace behaviour up with a custom type's validation and other behaviours.
+
+Without knowing how custom types will be used by provider developers, this option seems premature, and makes less sense than bundling validation functions with custom types.
```

Comment:
I think bundling a `ModifyPlan` with the type makes sense here. `RequiresReplaceIf` less so. For example, if we're implementing a timestamp resource, a `ModifyPlan` method would allow it to squash semantically meaningless diffs, rather than making provider developers do that everywhere the type is used.